File: app/src/main/java/com/kode/utair/domain/interactors/MainInteractor.java
package com.kode.utair.domain.interactors;

import com.kode.utair.domain.models.WeatherSettings;
import com.kode.utair.domain.usecases.GetCitiesUseCase;
import com.kode.utair.domain.usecases.GetWeatherSettingsUseCase;
import com.kode.utair.domain.usecases.SaveWeatherSettingsUseCase;
import com.kode.utair.domain.usecases.ValidateWeatherSettingsUseCase;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Single;

public class MainInteractor {

    private GetWeatherSettingsUseCase getWeatherForecastSettings;
    private SaveWeatherSettingsUseCase saveWeatherSettingsUseCase;
    private GetCitiesUseCase getCitiesUseCase;
    private ValidateWeatherSettingsUseCase validateWeatherSettingsUseCase;

    @Inject
    public MainInteractor(GetWeatherSettingsUseCase getWeatherForecastSettings,
                          SaveWeatherSettingsUseCase saveWeatherSettingsUseCase,
                          GetCitiesUseCase getCitiesUseCase,
                          ValidateWeatherSettingsUseCase validateWeatherSettingsUseCase) {
        this.getWeatherForecastSettings = getWeatherForecastSettings;
        this.saveWeatherSettingsUseCase = saveWeatherSettingsUseCase;
        this.getCitiesUseCase = getCitiesUseCase;
        this.validateWeatherSettingsUseCase = validateWeatherSettingsUseCase;
    }

    public Single<WeatherSettings> getWeatherSettings() {
        return getWeatherForecastSettings.getSingle();
    }

    public Completable saveWeatherSettings(WeatherSettings weatherSettings) {
        return saveWeatherSettingsUseCase.getCompletable(weatherSettings);
    }

    public Single<List<String>> getCities() {
        return getCitiesUseCase.getSingle();
    }

    public Completable validateData(WeatherSettings weatherSettings) {
        return validateWeatherSettingsUseCase.getCompletable(weatherSettings);
    }
}

File: app/src/main/java/com/kode/utair/domain/interactors/MainInteractor.kt
package com.kode.utair.domain.interactors

import com.kode.utair.domain.models.WeatherSettings
import com.kode.utair.domain.repository.ICityRepository
import com.kode.utair.domain.repository.IWeatherSettingsRepository
import io.reactivex.Completable
import io.reactivex.Single
import javax.inject.Inject

class MainInteractor @Inject constructor(
        private val weatherSettingsRepository: IWeatherSettingsRepository,
        private val weatherSettingsValidator: WeatherSettingsValidator,
        private val cityRepository: ICityRepository
) {

    fun getWeatherSettings(): Single<WeatherSettings> {
        return weatherSettingsRepository.getWeatherSettings()
    }

    fun saveWeatherSettings(weatherSettings: WeatherSettings): Completable {
        return weatherSettingsRepository.saveWeatherSettings(weatherSettings)
    }

    fun getCities(): Single<List<String>> {
        return cityRepository.getCitiesList()
    }

    fun validateData(weatherSettings: WeatherSettings): Completable {
        return weatherSettingsValidator.validate(weatherSettings)
    }

}